File: HotelManagementSystem.Server/Models/Auth/ApiResponse.cs
namespace HotelManagementSystem.Server.Models.Auth;

public class ApiResponse<T>(bool success, string? message = null, T? data = default)
{
    public bool Success { get; set; } = success;
    public string? Message { get; set; } = message;
    public T? Data { get; set; } = data;

    public static ApiResponse<T> SuccessResponse(T data, string? message = "Request successful") =>
        new(true, message, data);

    public static ApiResponse<T> FailureResponse(string message) =>
        new(false, message, default);
}

File: hotelmanagementsystem.client/src/models/api-response.ts
export class ApiResponse<T> {
  success: boolean;
  message?: string;
  data?: T;

  constructor(success: boolean, message?: string, data?: T) {
    this.success = success;
    this.message = message;
    this.data = data;
  }

  static successResponse<T>(data: T, message: string = "Request successful"): ApiResponse<T> {
    return new ApiResponse<T>(true, message, data);
  }

  static failureResponse<T>(message: string): ApiResponse<T> {
    return new ApiResponse<T>(false, message);
  }
}

export class PaginatedResponse<T> extends ApiResponse<T[]> {
  page: number;
  pageSize: number;
  totalCount: number;
  totalPages: number;

  constructor(data: T[], page: number, pageSize: number, totalCount: number, message: string = "Request successful") {
    super(true, message, data);
    this.page = page;
    this.pageSize = pageSize;
    this.totalCount = totalCount;
    this.totalPages = Math.ceil(totalCount / pageSize);
  }
}

